// @astrojs/language-server/dist/core/parseCSS.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractStylesheets = void 0;
const muggle_string_1 = require("muggle-string");
const buildMappings_js_1 = require("../buildMappings.js");
function extractStylesheets(styles) {
    return mergeCSSContexts(styles);
}
exports.extractStylesheets = extractStylesheets;
function mergeCSSContexts(inlineStyles) {
    const codes = [];
    for (const javascriptContext of inlineStyles) {
        if (javascriptContext.type === 'style-attribute')
            codes.push('__ { ');
        codes.push([
            javascriptContext.content,
            undefined,
            javascriptContext.position.start,
            {
                verification: true,
                completion: true,
                semantic: true,
                navigation: true,
                structure: true,
                format: false,
            },
        ]);
        if (javascriptContext.type === 'style-attribute')
            codes.push(' }\n');
    }
    const mappings = (0, buildMappings_js_1.buildMappings)(codes);
    const text = (0, muggle_string_1.toString)(codes);
    return {
        id: 'style.css',
        languageId: 'css',
        snapshot: {
            getText: (start, end) => text.substring(start, end),
            getLength: () => text.length,
            getChangeRange: () => undefined,
        },
        embeddedCodes: [],
        mappings,
    };
}
